File: src/main/java/xyz/emptydreams/mi/api/net/NetworkLoader.java
package xyz.emptydreams.mi.api.net;

import net.minecraftforge.fml.common.network.NetworkRegistry;
import net.minecraftforge.fml.common.network.simpleimpl.IMessage;
import net.minecraftforge.fml.common.network.simpleimpl.IMessageHandler;
import net.minecraftforge.fml.common.network.simpleimpl.SimpleNetworkWrapper;
import net.minecraftforge.fml.relauncher.Side;
import xyz.emptydreams.mi.ModernIndustry;
import xyz.emptydreams.mi.api.net.handler.ClientHandler;
import xyz.emptydreams.mi.api.net.handler.CommonMessage;
import xyz.emptydreams.mi.api.net.handler.ServerHandler;
import xyz.emptydreams.mi.api.register.others.AutoLoader;

import static net.minecraftforge.fml.relauncher.Side.CLIENT;
import static net.minecraftforge.fml.relauncher.Side.SERVER;

/**
 * 网络信息传递总注册器
 * @author EmptyDremas
 */
@AutoLoader
public final class NetworkLoader {

	private static final SimpleNetworkWrapper instance =
			NetworkRegistry.INSTANCE.newSimpleChannel(ModernIndustry.MODID);
	
	public static SimpleNetworkWrapper instance() {
		return instance;
	}
	
	/** 存储当前ID分配位点 */
	private static int nextID = -1;
	
	static {
		registerMessage(ServerHandler.class, CommonMessage.class, SERVER);
		registerMessage(ClientHandler.class, CommonMessage.class, CLIENT);
	}
	
	private static <REQ extends IMessage, REPLY extends IMessage> void registerMessage(
			Class<? extends IMessageHandler<REQ, REPLY>> messageHandler,
			Class<REQ> requestMessageType, Side side) {
		instance.registerMessage(messageHandler, requestMessageType, ++nextID, side);
	}
	
	private NetworkLoader() {
		throw new AssertionError("不应该调用该构造函数");
	}
	
}

File: src/main/java/top/kmar/mi/api/newnet/NetworkLoader.kt
package top.kmar.mi.api.newnet

import net.minecraftforge.fml.common.network.NetworkRegistry
import net.minecraftforge.fml.common.network.simpleimpl.IMessage
import net.minecraftforge.fml.common.network.simpleimpl.IMessageHandler
import net.minecraftforge.fml.common.network.simpleimpl.SimpleNetworkWrapper
import net.minecraftforge.fml.relauncher.Side
import top.kmar.mi.ModernIndustry
import top.kmar.mi.api.newnet.handlers.ClientHandler
import top.kmar.mi.api.newnet.handlers.ServerHandler
import top.kmar.mi.api.regedits.others.AutoLoader
import java.util.concurrent.atomic.AtomicInteger

/**
 * 自动化网络通信的 loader
 * @author EmptyDreams
 */
@AutoLoader
object NetworkLoader {

    @JvmStatic
    val instance: SimpleNetworkWrapper = NetworkRegistry.INSTANCE.newSimpleChannel(ModernIndustry.MODID)

    init {
        registerMessage(ServerHandler::class.java, CommonMessage::class.java, Side.SERVER)
        registerMessage(ClientHandler::class.java, CommonMessage::class.java, Side.CLIENT)
    }

    private val idIndex = AtomicInteger(0)

    private fun <REQ : IMessage?, REPLY : IMessage?> registerMessage(
        messageHandler: Class<out IMessageHandler<REQ, REPLY>?>,
        requestMessageType: Class<REQ>, side: Side
    ) {
        instance.registerMessage(messageHandler, requestMessageType, idIndex.getAndIncrement(), side)
    }

}